### PR TITLE
Appendix g dev debug unmet load hour issues

### PR DESCRIPTION
**Pull request by weilixu**
Bug fixes including
1. Add increment values that breaks the unmet load hour break loop
2. Add statement to close SQL sessions before running a new simulation in the same folder (issues under Window OS)
3. Add error handler when simulation failed in the unmet load hour loop.
4. Add schedule type check for optimal start EMS code

### CODE REVIEW

File: lib/openstudio-standards/standards/Standards.Model.rb [hunk 1]
```diff
@@ -455,66 +455,74 @@ def model_create_prm_any_baseline_building(user_model, building_type, climate_zo
...
+            OpenStudio.logFree(OpenStudio::Error, 'openstudio.standards.Model', "After 8 rounds of zone sizing factor adjustments the unmet load hours for the baseline model (#{degs} degree of rotation) still exceed 300 hours. Please open an issue on GitHub (https://github.com/NREL/openstudio-standards/issues) and share your user model with the developers.")
+            break
+          end
+          # Close the previous SQL session if any - this could prevent EnergyPlus overloading the same session
+          sql = model.sqlFile.get
+          if sql.connectionOpen
+            sql.close
+          end
+          if model_run_simulation_and_log_errors(model, "#{sizing_run_dir}/final#{degs}")
+            # model_run_simulation_and_log_errors(model, "#{sizing_run_dir}/final#{degs}") == false
```
Comment (lymereJ) [inline lib/openstudio-standards/standards/Standards.Model.rb:469]: Maybe we should remove that line.

File: lib/openstudio-standards/standards/Standards.Model.rb [hunk 2]
```diff
@@ -455,66 +455,74 @@ def model_create_prm_any_baseline_building(user_model, building_type, climate_zo
...
+                  # Make adjustment to zone heating sizing factor
+                  # Do not adjust factors greater or equal to 2
+                  # Rewrite the logic to avoid 5 level nested logic.
+                  htg_umlh > 150 ? sizing_factor = [2.0, sizing_factor * 1.1].min : sizing_factor = [2.0, sizing_factor * 1.05].min
+                  # if sizing_factor < 2.0
+                  #  if htg_umlh > 150
+                  #    sizing_factor *= 1.1
+                  #  else #elsif htg_umlh > 50 is not needed for the same reason
+                  #    sizing_factor *= 1.05
+                  #  end
```
Comment (lymereJ) [inline lib/openstudio-standards/standards/Standards.Model.rb:515]: Maybe we should remove these comments?

File: lib/openstudio-standards/standards/Standards.Model.rb [hunk 3]
```diff
@@ -455,66 +455,74 @@ def model_create_prm_any_baseline_building(user_model, building_type, climate_zo
...
                   end
+                  # Make adjustment to zone cooling sizing factor
+                  # Do not adjust factors greater or equal to 2
+                  # Rewrite the logic to avoid 5 level nested logic.
+                  clg_umlh > 150 ? sizing_factor = [2.0, sizing_factor * 1.1].min : sizing_factor = [2.0, sizing_factor * 1.05].min
+                  # if clg_umlh > 150
+                  #  sizing_factor *= 1.1
+                  # else # elseif clg_umlh > 50 is not needed since the code block only reached when clg_umlh>50
+                  #  sizing_factor *= 1.05
+                  # end
```
Comment (lymereJ) [inline lib/openstudio-standards/standards/Standards.Model.rb:492]: Maybe we should remove these comments?

File: lib/openstudio-standards/standards/Standards.Model.rb [hunk 4]
```diff
@@ -455,66 +455,74 @@ def model_create_prm_any_baseline_building(user_model, building_type, climate_zo
...
-                  if clg_umlh > 150
-                    sizing_factor *= 1.1
-                  elsif clg_umlh > 50
-                    sizing_factor *= 1.05
+                  if thermal_zone.sizingZone.zoneCoolingSizingFactor.is_initialized
+                    sizing_factor = thermal_zone.sizingZone.zoneCoolingSizingFactor.get
                   end
+                  # Make adjustment to zone cooling sizing factor
+                  # Do not adjust factors greater or equal to 2
+                  # Rewrite the logic to avoid 5 level nested logic.
```
Comment (dmaddoxwhite) [inline lib/openstudio-standards/standards/Standards.Model.rb:486]: Don't need to keep "Rewrite... " comment.

File: lib/openstudio-standards/standards/Standards.Model.rb [hunk 5]
```diff
@@ -455,66 +455,74 @@ def model_create_prm_any_baseline_building(user_model, building_type, climate_zo
...
-                  elsif htg_umlh > 50
-                    sizing_factor *= 1.05
+                  if thermal_zone.sizingZone.zoneHeatingSizingFactor.is_initialized
+                    # Get zone heating sizing factor
+                    sizing_factor = thermal_zone.sizingZone.zoneHeatingSizingFactor.get
                   end
+
+                  # Make adjustment to zone heating sizing factor
+                  # Do not adjust factors greater or equal to 2
+                  # Rewrite the logic to avoid 5 level nested logic.
```
Comment (dmaddoxwhite) [inline lib/openstudio-standards/standards/Standards.Model.rb:508]: Remove "Rewrite ... " comment.

**Review by dmaddoxwhite — APPROVED** [review]

**Comment by mdahlhausen** [comment]
@weilixu @lymereJ 8 is a lot of adjustments; that's a lot of time a user will be at their PC while the measure is looping through sizing runs.  I have a commit drafted to set it to three instead.  Is that ok?

**Comment by lymereJ** [comment]
> @weilixu @lymereJ 8 is a lot of adjustments; that's a lot of time a user will be at their PC while the measure is looping through sizing runs. I have a commit drafted to set it to three instead. Is that ok?

@dmaddoxwhite, is that okay?

**Comment by weilixu** [comment]
> s; that's a lot of time a user will be at their PC while the measure is looping through sizing runs. I have a commit drafted to set it to three instead. Is that ok?

I am ok with it, I wonder if we should bump up the multiplier to account for the reduced iterations.

**Comment by lymereJ** [comment]
The multipliers are from the PRM-RM. So we probably would want to specify that we changed the approach if that's what we want to do.

**Comment by dmaddoxwhite** [comment]
How about adding an argument to the method to set the maximum number of sizing factor adjustment runs? If we limit to three, without an override, a user might get a model that fails due to unmet hours with no way to fix it.

**Comment by mdahlhausen** [comment]
@dmaddoxwhite in my experience, equipment size is rarely if ever the source of unmet hours.  If one iteration doesn't fix it, eight iterations won't either.  The issue is more likely to do with schedule alignment (e.g. 5% occupancy when the HVAC system is turned off), design conditions (extremes in the annual weather file very different from design conditions), or some weird VAV sizing issue.  For many of these, even an HVAC systems with 10x the capacity isn't going to solve them.

Users will generate models that have baselines that fail the UMLH test.  I'd rather they fail faster (3 iterations) than wait for 8 and wonder why it still didn't work.  Alternatively, we can add some QAQC checks to point out basic issues (like 5% occupancy in zones when the HVAC system is off).  I've already got methods in standards to check for those sorts of things.

**Review by mdahlhausen — APPROVED** [review]
looks fine; made some minor cleanup fixes and changed number of adjustments to 3.  It may be worth adding some checking in here to pick out the cause of unmet hours, rather than just increasing sizing factors.  The PRM RM includes some direction beyond sizing, but IMO it's not very specific or actionable.

**Comment by mdahlhausen** [comment]
@weilixu @lymereJ @dmaddoxwhite good to merge.  Up to you if you want to edit the # of adjustments further or make that a user argument.  It's also fine as is.

**Comment by weilixu** [comment]
> @weilixu @lymereJ @dmaddoxwhite good to merge. Up to you if you want to edit the # of adjustments further or make that a user argument. It's also fine as is.

Thanks @mdahlhausen, we will update the number of iteration.
We also need to disable an optimal control in the PRM method. Will make a push shortly.

**Pull request merged**